registry.go

```diff
@@ -1,20 +1,36 @@
 package distribution
 
-import (
-	"io"
-	"time"
+import "golang.org/x/net/context"
 
-	"github.com/docker/distribution/digest"
-	"github.com/docker/distribution/manifest"
-	"golang.org/x/net/context"
-)
+// Scope defines the set of items that match a namespace.
+type Scope interface {
+	// Contains returns true if the name belongs to the namespace.
+	Contains(name string) (bool, error)
+}
+
+// Namespace represents a collection of repositories, addressable by name.
+// Generally, a namespace is backed by a set of one or more services,
+// providing facilities such as registry access, trust, and indexing.
+type Namespace interface {
+	// Scope describes the names that can be used with this Namespace. The
+	// global namespace will have a scope that matches all names. The scope
+	// effectively provides an identity for the namespace.
+	Scope() Scope
 
-// Registry represents a collection of repositories, addressable by name.
-type Registry interface {
 	// Repository should return a reference to the named repository. The
 	// registry may or may not have the repository but should always return a
 	// reference.
 	Repository(ctx context.Context, name string) (Repository, error)
+
+	// TODO(stevvooe): Namespaces should provide other services that allow one
+	// to explore the contents of the namespace. One such example is an
+	// "index" or search service. A commented example is provided here to
+	// illustrate the concept.
+	//
+	// Index returns an IndexService that provides search across the
+	// namespace, supporting content discovery and collective information
+	// about available images.
+	// Index(ctx context.Context) (IndexService, error)
```

Comment:
We should take this into consideration now since we are defining the registry service as an implementation of Namespace.  If namespace covers more than repositories (which I think we agree it does), then what does it mean for Registry to be a partial implementation?  Should a registry throw an error when an Index is retrieved, or should Namespace be a composed type.